Google.Ads.Gax/src/Util/ExceptionUtilities.cs

```diff
@@ -0,0 +1,197 @@
+        /// <returns>The parsed <see cref="AdsBaseException"/> if parsing is successful;
+        /// The underlying <see cref="RpcException" /> if the exception cannot be parsed as a
+        /// AdsBaseException.</returns>
+        /// <remarks><code>AggregateException</code> is very close to a catch-all exception for
+        /// Tasks. While all the situations that we know of involves this exception being thrown
+        /// by an underlying <code>RpcException</code>, theoretically this method may return a
+        /// null object, if the <code>AggregateException</code> is not caused by an
+        /// <code>RpcException</code>. That would typically indicate an underlying issue with
+        /// the code.</remarks>
+        internal static RpcException ParseTaskException<TResponse>(AggregateException e)
+        {
+            RpcException rpcException = ExtractRpcException(e);
+            AdsBaseException adsException = ParseRpcException<TResponse>(rpcException);
+            return (adsException == null) ? rpcException : adsException;
+        }
+
+        /// <summary>
+        /// Extracts the RPC exception from an <see cref="AggregateException"/>.
+        /// </summary>
+        /// <param name="aggregateException">The aggregate exception.</param>
+        /// <returns>the extracted <see cref="RpcException"/>, or null if an exception
+        /// cannot be extracted.</returns>
+        internal static RpcException ExtractRpcException(AggregateException aggregateException)
+        {
+            return aggregateException?.InnerExceptions?.FirstOrDefault(
+                delegate (Exception innerException)
+                {
+                    return (innerException is RpcException);
+                }
+            ) as RpcException;
+        }
+
+        /// <summary>
+        /// Parses the RPC exception into a <see cref="AdsBaseException" />.
+        /// </summary>
+        /// <param name="rpcException">The RPC exception.</param>
+        /// <returns>The parsed exception, or a null if the parsing cannot be done.</returns>
+        internal static AdsBaseException ParseRpcException<TResponse>(RpcException rpcException)
+        {
+            if (rpcException == null)
+            {
+                return null;
+            }
+
+            System.Type exceptionType = FindAdsExceptionType<TResponse>(rpcException);
+
+            if (exceptionType == null)
+            {
+                return null;
+            }
+            MethodInfo methodInfo = exceptionType.GetMethod("Create",
+                BindingFlags.Public | BindingFlags.Static);
+            if (methodInfo == null)
+            {
+                return null;
+            }
+            return methodInfo.Invoke(null, new object[] { rpcException })
+                as AdsBaseException;
+        }
+
+        /// <summary>
+        /// Finds the type of the ads exception that matches <paramref name="rpcException"/>.
```

Comment:
The wording of this made me think this was finding the inner exception that matches the *type* of `rpcException`, but this is actually inspecting the `rpcException` and returning the *type* of Ads exception that `rpcException` can be interpreted as, correct?

Reply:
Yep, this is correct. I rephrased the comment.